gimp/pictonode-gimp-plugin/GIMP 2.99/pictonode/custom_nodes.py

```diff
@@ -252,80 +318,442 @@ def __init__(self, node_window, *args, **kwds) -> None:
+        self.value_update()
+
+    def process_ouput(self):
+        '''
+        Updates buffer reference in map
+        '''
+
+        print("Buffer: ", self.buffer)
+        self.node_window.buffer_map[self.buffer_id] = [self.buffer, self.layer]
+
+        if self.buffer:
+            return True
+        return False
+
+    def value_update(self):
+        '''
+        Processes image and sends out updated buffer reference
+        '''
+
+        did_process = self.process_ouput()
+
+        if not did_process:
+            print("Error: could not process composite")
+
+        self.node_socket_output.write(bytes(self.buffer_id, 'utf8'))
+
+    def node_socket_disconnect(self, socket, sink, datasource):
+        '''
+        Processes socket disconnect and updates output accordingly
+        '''
+
+        # reset buffer reference information
+        if datasource == 1:
+            self.incoming_buffer1_id = None
+            self.incoming_buffer1 = None
+        elif datasource == 2:
+            self.incoming_buffer2_id = None
+            self.incoming_buffer2 = None
+
+        self.buffer = None
+        self.layer = None
+        self.process_input()
+
+    def node_socket_connect(self, sink, source):
+        '''
+        Sends buffer reference upon initial socket connection
+        '''
+
+        self.value_update()
+
+    def node_socket_incoming(self, socket, payload, datasource):
+        '''
+        Updates node internal state upon update from socket.
+        '''
+
+        # reset buffer reference based on which socket is changed
+        if datasource == 1:
+            self.incoming_buffer1_id = None
+            self.incoming_buffer1 = None
+        elif datasource == 2:
+            self.incoming_buffer2_id = None
+            self.incoming_buffer2 = None
+
+        self.buffer = None
+        self.layer = None
+
+        # set new buffer information
+        if payload:
+            self.image_context.reset_context()
+            if datasource == 1:
+                self.incoming_buffer1_id = payload.decode('utf-8')
+                print("Buffer ID 1 incoming: ", self.incoming_buffer1_id)
+            elif datasource == 2:
+                self.incoming_buffer2_id = payload.decode('utf-8')
+                print("Buffer ID 2 incoming: ", self.incoming_buffer2_id)
+
+            # Grab or regrab each buffer from buffer map
+            self.incoming_buffer1 = self.node_window.buffer_map.get(
+                self.incoming_buffer1_id)[0]
+
+            self.incoming_buffer2 = self.node_window.buffer_map.get(
+                self.incoming_buffer2_id)[0]
+
+            self.layer = self.node_window.buffer_map.get(
+                self.incoming_buffer1_id)[1]
+
+            print("Comp Incoming 1: ", self.incoming_buffer1)
+            print("Comp Incoming 2: ", self.incoming_buffer2)
+
+            self.process_input()
+        else:
+            print("Error!!!")
+
+
+class BlurNode(GtkNodes.Node):
```

Comment:
There seems to be a lot of duplicated logic between these custom nodes. Maybe it would be best to have an `OntarioNode` parent class that abstracts over some of this, and takes the Ontario function, the number/type of inputs and the number/type of outputs as a constructor?